### PR TITLE
Snap updated to 3.3.3

### PR DESCRIPTION
**Pull request by sklassen**
## Overview

Updated README-SNAP.md (removed process monitor; replace passwords, uuid and secrets with environment variables; showed bash scripts need to generate passwords, uuids and secrets.)

Updated hook scripts. configure now uses sed in-place mode; install now supports multiple instances; post refresh is a copy of install, possibly needed for those that transition from 2.x to 3.x

Rewrote snapcraft yaml to use version 3.3.2 from the tar release. 

.gitignore now also ignores all .snap files.

## Testing recommendations

I've release this version to couchdb-sklassen. 

## GitHub issue number

Fixes apache/couchdb-pkg#105

## Checklist

- [ X ] Code is written and works correctly;
- [ X ] Changes are covered by tests;
- [ X ] Documentation reflects the changes;

### CODE REVIEW

File: README-SNAP.md [hunk 1]
```diff
@@ -251,30 +277,30 @@ you are using the default.
...
 
 ```bash
-sudo snap install ./couchdb_3.0.0_amd64.snap --dangerous
+> lxc launch ubuntu-daily:22.04 cdb
+> lxc shell cdb
+$ snapcraft --destructive-mode --verbose
 ```
 
-Clean up with:
+Once the snap has been built, the snap can be installed locally using `--dangerous`:
```
Comment (big-r81) [inline README-SNAP.md:301]: this `--dangerous` mode is only for testing local snaps? When installing from the official "Snap repos" you would install couchdb without this flag?

File: README-SNAP.md [hunk 2]
```diff
@@ -251,30 +277,30 @@ you are using the default.
...
 
 ```bash
-$ sudo snapcraft clean          # sudo may not be required but is safe
+sudo snap install ./couchdb_3.2.2_amd64.snap --dangerous
```
Comment (big-r81) [inline README-SNAP.md:304]: use 3.3.2 as version

File: snap/snapcraft.yaml [hunk 3]
```diff
@@ -1,106 +1,77 @@
...
-    plugin: dump
-    source: https://apache.bintray.com/couchdb-deb/pool/C/CouchDB/couchdb_3.1.1~focal_amd64.deb
-    source-type: deb
-    # because this doesn't use apt, we have to manually list all of our
-    # dependencies :(
+    plugin: make
+    source: https://dlcdn.apache.org/couchdb/source/3.3.2/apache-couchdb-3.3.2.tar.gz
+    source-type: tar
+    override-build: |
+            ./configure --disable-docs --spidermonkey-version 91
```
Comment (big-r81) [inline snap/snapcraft.yaml:22]: Why did you disabled the inline docs?

File: README-SNAP.md (hunk 1 above)
Comment (sklassen) [reply README-SNAP.md:301]: Correct. --dangerous is only needed when installing a snap that you've rolled your own. And even then, I don't think the adjective is apt: the snap is still embedded within AppAmour. Snap that are uploaded to the store have an extra check to see that they aren't tainted by external links.

File: README-SNAP.md (hunk 2 above)
Comment (sklassen) [reply README-SNAP.md]: Rolled version number in the example.

File: snap/snapcraft.yaml (hunk 3 above)
Comment (sklassen) [reply snap/snapcraft.yaml:22]: Speed of building and smaller snap size. Probably more of the former than the latter.

File: README-SNAP.md [hunk 4]
```diff
@@ -56,9 +67,15 @@ Erlang settings are stored in the `/var/snap/couchdb/current/etc/vm.args` file.
...
 ```bash
-$ sudo snap set couchdb name=couchdb@1.2.3.4 setcookie=cutter
+$ sudo snap set couchdb name=couchdb@1.2.3.4 setcookie=$COOKIE
 ```
 
+Where COOKIE is an enviroment variable. You can auto generated a cookie with the command 
+below. 
+
+```bash
+$ export COOKIE=`echo $RANDOM | md5sum | head -c 20; echo;`
```
Comment (nickva) [inline README-SNAP.md:77]: This seems to work in `bash` though in other shells like `zsh` it seems to always generate the same value which might be dangerous:

```
% export COOKIE=`echo $RANDOM | md5sum | head -c 20; echo;`
% echo $COOKIE
dc8b85e421d60214e049
% export COOKIE=`echo $RANDOM | md5sum | head -c 20; echo;`
% echo $COOKIE
dc8b85e421d60214e049
```

In couchdb-pkg we use something like:

```
echo $(dd if=/dev/random bs=1 count=38 status=none | base64 | tr -cd '[:alnum:]')
9qiRUeH4SMYCNoP8emESCrhMrQtf6q0EKv98s6McMfUvS33r7Y
```

Comment (sklassen) [reply README-SNAP.md]: Ok. I modified to use `/dev/random` to bake a better cookie

**Comment by big-r81** [comment]
Hey Simon @sklassen, i tried the `snapcraft remote-build` tool, but it fails. I uploaded the log file. I don't know if it's a problem of the remote build or with the snap .
[couchdb_amd64.1.txt](https://github.com/apache/couchdb-pkg/files/13325738/couchdb_amd64.1.txt)

**Comment by sklassen** [comment]
> Hey Simon @sklassen, i tried the `snapcraft remote-build` tool, but it fails. I uploaded the log file. I don't know if it's a problem of the remote build or with the snap . [couchdb_amd64.1.txt](https://github.com/apache/couchdb-pkg/files/13325738/couchdb_amd64.1.txt)

I ran and also saw the same error. It did worked on a previous version of snapcraft. The arguments verbose has also changed since last I ran it. Checking the logs and comparing to previous runs, I can see it executing `configure` but not `make`. It worked when I hand ran the override in the build directory and then manually ran `snapcraft prime`

Maybe something with newline. 

`override-build: ./configure --disable-docs --spidermonkey-version 91 && make release && cp -r rel/couchdb/* $SNAPCRAFT_PART_INSTALL`

This worked for me locally and when running remote-build. This is the snapcraft location and version I used.

```
/snap/bin/snapcraft
root@cdb22:~/src/github.com/sklassen/couchdb-pkg# snapcraft --version
snapcraft 7.5.3   
```
After which

```
Build status as of 2023-11-13 18:17:58.993147:
	arch=amd64	state=Successfully built
	arch=arm64	state=Successfully built
Snapped couchdb_3.3.2_amd64.snap
Build log available at 'couchdb_amd64.1.txt'
Snapped couchdb_3.3.2_arm64.snap
Build log available at 'couchdb_arm64.1.txt'
Build complete.
```
Try again and see if it works for you. You should then see a `couchdb_3.3.2_amd64.snap` and `couchdb_3.3.2_arm64.snap`.

It would be better if we didn't have to have an override-build. It is needed because i) couchdb configure doesn't have a --prefix (I believe it did in the 2.3.x branch but disappeared at some point) and ii) the Makefile is missing `install`. With these the snapcraft autoconf module should work without override.

**Comment by big-r81** [comment]
Hey Simon,

I tested your latest updates and it work for me. I tried to build the snaps for all other architectures too (except i386). Great work so far!

We are building binary packages (.deb) for Debian and Ubuntu distributions. Could we use this instead of building it again from source?

**Comment by sklassen** [comment]
> We are building binary packages (.deb) for Debian and Ubuntu distributions. Could we use this instead of building it again from source?

They are different paradigms and shoehorning an old file format into a new one seems perverse.

On the practical side, CouchDB .deb's /opt/ clashes within .snap's /snap/ and /var/snap structure. But the stronger case is on the principle side. Snaps security enhancement over .deb comes from the `snapcraft` build ensuring that the source is untainted by outside assets. By adding an build-override that copies binaries from -- official while it might be -- apache server goes against this spirit. The yaml file wants a source, I say give it the source.

**Comment by big-r81** [comment]
> > We are building binary packages (.deb) for Debian and Ubuntu distributions. Could we use this instead of building it again from source?
> 
> They are different paradigms and shoehorning an old file format into a new one seems perverse.
> 
> On the practical side, CouchDB .deb's /opt/ clashes within .snap's /snap/ and /var/snap structure. But the stronger case is on the principle side. Snaps security enhancement over .deb comes from the `snapcraft` build ensuring that the source is untainted by outside assets. By adding an build-override that copies binaries from -- official while it might be -- apache server goes against this spirit. The yaml file wants a source, I say give it the source.

Ok, got it.

One question to the Base snap. Can a `core22` base run on an Ubuntu 18.04 or does this base restrict to a special version?

**Comment by sklassen** [comment]
> One question to the Base snap. Can a `core22` base run on an Ubuntu 18.04 or does this base restrict to a special version?

It can. In the past I was running core20 on a Ubuntu 16.04 for the longest time. I think 16.04 is the oldest supported vintage. The same applies for core22, the snap while confining some system calls, abstracts permitted system calls between the snap and the base system. It's pretty useful for those edge cases with old hardware or dated OSs. (You know, like when you have a collection of unreliable common hardware that you want to network together).

**Review by big-r81 — APPROVED** [review]

**Pull request merged**